akka-http-caching/src/main/scala/akka/http/scaladsl/server/directives/CachingDirectives.scala

```diff
@@ -0,0 +1,91 @@
+
+import akka.actor.ActorRefFactory
+import akka.annotation.ApiMayChange
+import akka.http.caching.{ Cache, LfuCache }
+import akka.http.scaladsl.server.Directive0
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ ExecutionContext, Future }
+import akka.http.scaladsl.server._
+import akka.http.scaladsl.model.HttpMethods.GET
+import akka.http.scaladsl.model.headers._
+import akka.http.scaladsl.model.headers.CacheDirectives._
+import akka.http.scaladsl.server.RouteResult.Rejected
+
+@ApiMayChange
+trait CachingDirectives {
+  import akka.http.scaladsl.server.directives.BasicDirectives._
+  import akka.http.scaladsl.server.directives.RouteDirectives._
+
+  /**
+   * Wraps its inner Route with caching support using the given [[akka.http.caching.Cache]] implementation and
+   * the in-scope keyer function.
+   */
+  def cache(csm: CacheSpecMagnet): Directive0 = cachingProhibited | alwaysCache(csm)
+
+  /**
+   * Passes only requests to the inner route that explicitly forbid caching with a `Cache-Control` header with either
+   * a `no-cache` or `max-age=0` setting.
+   */
+  def cachingProhibited: Directive0 =
+    extract(_.request.headers.exists {
+      case x: `Cache-Control` ⇒ x.directives.exists {
+        case `no-cache`   ⇒ true
+        case `max-age`(0) ⇒ true
+        case _            ⇒ false
+      }
+      case _ ⇒ false
+    }).flatMap(if (_) pass else reject)
+
+  /**
+   * Wraps its inner Route with caching support using the given [[akka.http.caching.Cache]] implementation and
+   * in-scope keyer function. Note that routes producing streaming responses cannot be wrapped with this directive.
+   */
+  def alwaysCache(csm: CacheSpecMagnet): Directive0 = {
+    import csm._
+    mapInnerRoute { route ⇒ ctx ⇒
+      liftedKeyer(ctx) match {
+        case Some(key) ⇒ responseCache.apply(key, () ⇒ route(ctx))
+        case None      ⇒ route(ctx)
+      }
+    }
+  }
+
+  //# route-Cache
+  def routeCache(maxCapacity: Int = 500, initialCapacity: Int = 16, timeToLive: Duration = Duration.Inf,
+                 timeToIdle: Duration = Duration.Inf): Cache[RouteResult] = {
+    LfuCache(maxCapacity, initialCapacity, timeToLive, timeToIdle)
+  }
+  //#
+}
+
+object CachingDirectives extends CachingDirectives
+
+trait CacheSpecMagnet {
+  def responseCache: Cache[RouteResult]
+  def liftedKeyer: RequestContext ⇒ Option[Any]
+  implicit def executionContext: ExecutionContext
+}
+
+object CacheSpecMagnet {
+  implicit def apply(cache: Cache[RouteResult])(implicit keyer: CacheKeyer, factory: ActorRefFactory) = // # CacheSpecMagnet
+    new CacheSpecMagnet {
+      def responseCache = cache
+      def liftedKeyer = keyer.lift
+      implicit def executionContext = factory.dispatcher
+    }
+}
+
+trait CacheKeyer extends (PartialFunction[RequestContext, Any])
+
+object CacheKeyer {
+  implicit val Default: CacheKeyer = CacheKeyer {
+    case r: akka.http.scaladsl.server.RequestContext if r.request.method == GET ⇒ r.request.uri
```

Comment:
I know that's same as in spray but I think it's somewhat dangerous to only regard the URI as the key. If you place the cache at the wrong level of the routing hierarchy it might remove access controls, use cookie values from other users etc.

So we should at least try to be a bit more diligent to figure out what kind of headers to take into account as well (or use as a blacklist). E.g. we could say also never to cache response that contain `Authorization` headers. It's harder for cookies because we don't know which cookies are relevant for routing and result generation.

And then we need to figure out how to provide reasonable default behavior. E.g. we could also change authentication directives to add `no-cache` headers.

Or we decide to bail out and add a fat warning that users need to know what they are using when the `Default` keyer is used. Does someone has any information about how other software, e.g. transparent caches, deal with this problem?

Any other ideas?